src/core/dfn-panel.runtime.js

```diff
@@ -0,0 +1,168 @@
+        panel.remove();
+        break;
+      }
+    }
+  });
+}
+
+/** @param {HTMLElement} clickTarget */
+function deriveAction(clickTarget) {
+  const hitALink = !!clickTarget.closest("a");
+  if (clickTarget.closest("dfn, .index-term")) {
+    return hitALink ? null : "show";
+  }
+  if (clickTarget.closest("#dfn-panel")) {
+    if (hitALink) {
+      const clickedSelfLink = clickTarget.classList.contains("self-link");
+      return clickedSelfLink ? "hide" : "dock";
+    }
+    const panel = clickTarget.closest("#dfn-panel");
+    return panel.classList.contains("docked") ? "hide" : null;
+  }
+  if (document.getElementById("dfn-panel")) {
+    return "hide";
+  }
+  return null;
+}
+
+/** @param {HTMLElement} dfn */
+function createPanel(dfn) {
+  const { id } = dfn;
+  const href = dfn.dataset.href || `#${id}`;
+  const links = document.querySelectorAll(`a[href="${href}"]:not(.index-term)`);
+
+  const tempNode = document.createElement("div");
+  tempNode.innerHTML = `
+    <aside class="dfn-panel" id="dfn-panel">
+      <b><a class="self-link" href="${href}">Permalink</a></b>
+      <b>Referenced in:</b>
+      ${referencesToHTML(id, links)}
+    </aside>
+  `;
+  return tempNode.querySelector("aside");
+}
+
+/**
+ * @param {string} id dfn id
+ * @param {NodeListOf<HTMLLinkElement>} links
+ * @returns {HTMLUListElement}
+ */
+function referencesToHTML(id, links) {
+  if (!links.length) {
+    return `<ul><li>Not referenced in this document.</li></ul>`;
+  }
+
+  /** @type {Map<string, string[]>} */
+  const titleToIDs = new Map();
+  links.forEach((link, i) => {
+    const linkID = link.id || `ref-for-${id}-${i + 1}`;
+    if (!link.id) link.id = linkID;
+    const title = getReferenceTitle(link);
+    const ids = titleToIDs.get(title) || titleToIDs.set(title, []).get(title);
+    ids.push(linkID);
+  });
+
+  /**
+   * Returns a list that is easier to render in `listItemToHTML`.
+   * @param {[string, string[]]} entry an entry from `titleToIDs`
+   * @returns {{ title: string, id: string }[]} The first list item contains
+   * title from `getReferenceTitle`, rest of items contain strings like `(2)`,
+   * `(3)` as title.
+   */
+  const toLinkProps = ([title, ids]) => {
+    return [{ title, id: ids[0] }].concat(
+      ids.slice(1).map((id, i) => ({ title: `(${i + 2})`, id }))
+    );
+  };
+
+  /**
+   * @param {[string, string[]]} entry
+   * @returns {HTMLLIElement}
+   */
+  const listItemToHTML = entry =>
+    `<li>${toLinkProps(entry)
+      .map(link => `<a href="#${link.id}">${link.title}</a>${" "}`)
+      .join("")}</li>`;
+
+  const listItems = [...titleToIDs].map(listItemToHTML);
+  return `<ul>${listItems.join("")}</ul>`;
+}
+
+/** @param {HTMLAnchorElement} link */
+function getReferenceTitle(link) {
+  const section = link.closest("section");
+  if (!section) return null;
+  const heading = section.querySelector("h1, h2, h3, h4, h5, h6");
+  if (!heading) return null;
+  return norm(heading.textContent);
```

Comment:
I think we can use `.innerText.trim()` here which does the normalization for us. Or we can just inline it as no other line uses it.